File: lib/domain/shared/dtos/session_principal.rb
# frozen_string_literal: true

module Domain
  module Shared
    module Dtos
      # セッション Cookie または API キーから解決した「ログイン主体」。
      # ActiveRecord の User はアダプタ境界に閉じ、ドメイン・コントローラ・ビューは本 DTO のみを扱う。
      class SessionPrincipal
        attr_reader :id, :email, :name, :admin

        def initialize(id:, email:, name:, admin:, anonymous:)
          @id = id
          @email = email
          @name = name
          @admin = admin
          @anonymous = anonymous
        end

        def admin?
          @admin
        end

        def anonymous?
          @anonymous
        end

        def authenticated?
          !anonymous?
        end
      end
    end
  end
end

File: crates/agrr-domain/src/shared/dtos/session_principal.rs
/// Ruby: `Domain::Shared::Dtos::SessionPrincipal`
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SessionPrincipal {
    pub id: i64,
    pub email: String,
    pub name: String,
    pub admin: bool,
    pub anonymous: bool,
    /// `Some` when resolved via API key; `None` for session cookie (unrestricted Masters access).
    pub api_key_scopes: Option<Vec<String>>,
}

impl SessionPrincipal {
    pub fn admin(&self) -> bool {
        self.admin
    }

    pub fn anonymous(&self) -> bool {
        self.anonymous
    }

    pub fn authenticated(&self) -> bool {
        !self.anonymous
    }
}
